File: tests/harness/lib/assertions.ts
/**
 * Filesystem and interview-specific assertions.
 */

import { existsSync, readFileSync, readdirSync, statSync } from "fs";
import { join } from "path";

// ── Filesystem assertions ───────────────────────────────────────────

export function fileExists(dir: string, relativePath: string): boolean {
  return existsSync(join(dir, relativePath));
}

export function fileContains(
  dir: string,
  relativePath: string,
  substring: string
): boolean {
  const fullPath = join(dir, relativePath);
  if (!existsSync(fullPath)) return false;
  return readFileSync(fullPath, "utf-8").includes(substring);
}

export function fileMatches(
  dir: string,
  relativePath: string,
  pattern: RegExp
): boolean {
  const fullPath = join(dir, relativePath);
  if (!existsSync(fullPath)) return false;
  return pattern.test(readFileSync(fullPath, "utf-8"));
}

export function readFile(dir: string, relativePath: string): string {
  return readFileSync(join(dir, relativePath), "utf-8");
}

export function listFiles(dir: string, relativePath: string): string[] {
  const fullPath = join(dir, relativePath);
  if (!existsSync(fullPath)) return [];
  return readdirSync(fullPath)
    .filter((f) => statSync(join(fullPath, f)).isFile())
    .sort();
}

export function listDirs(dir: string, relativePath: string): string[] {
  const fullPath = join(dir, relativePath);
  if (!existsSync(fullPath)) return [];
  return readdirSync(fullPath)
    .filter((f) => statSync(join(fullPath, f)).isDirectory())
    .sort();
}

// ── Interview-specific assertions ───────────────────────────────────

/** Timestamp filename pattern: YYYY-MM-DD-HH-MM-SS-slug.md */
const TIMESTAMP_FILENAME = /^\d{4}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}-.+\.md$/;

/**
 * Verify all files in a directory match the timestamp naming convention.
 */
export function allFilesMatchTimestampPattern(
  dir: string,
  relativePath: string
): boolean {
  const files = listFiles(dir, relativePath);
  return files.length > 0 && files.every((f) => TIMESTAMP_FILENAME.test(f));
}

/**
 * Verify a file has valid YAML frontmatter with required interview fields.
 */
export function hasValidFrontmatter(
  dir: string,
  relativePath: string,
  requiredFields: string[]
): boolean {
  const content = readFile(dir, relativePath);
  const match = content.match(/^---\n([\s\S]*?)\n---/);
  if (!match) return false;

  const frontmatter = match[1];
  return requiredFields.every((field) =>
    new RegExp(`^${field}:`, "m").test(frontmatter)
  );
}

/**
 * Parse the test log (JSONL) and return events.
 */
export interface TestLogEvent {
  event: string;
  specialist?: string;
  mode?: string;
  question?: string;
  transcript_file?: string;
  analysis_file?: string;
  transcript_id?: string;
  topic?: string;
  action?: string;
  exchanges?: number;
  specialists_invoked?: string[];
  files_written?: number;
  timestamp: string;
}

export function parseTestLog(
  dir: string,
  relativePath: string
): TestLogEvent[] {
  const content = readFile(dir, relativePath);
  return content
    .trim()
    .split("\n")
    .filter((line) => line.trim())
    .map((line) => JSON.parse(line));
}

/**
 * Extract the list of specialists invoked from the test log.
 */
export function specialistsInvoked(events: TestLogEvent[]): string[] {
  return [
    ...new Set(
      events
        .filter((e) => e.event === "specialist_invoked")
        .map((e) => e.specialist!)
    ),
  ];
}

/**
 * Count exchanges (question_asked events) in the test log.
 */
export function exchangeCount(events: TestLogEvent[]): number {
  return events.filter((e) => e.event === "question_asked").length;
}

/**
 * Verify every question_asked has a matching answer_received.
 */
export function allQuestionsAnswered(events: TestLogEvent[]): boolean {
  const questions = events.filter((e) => e.event === "question_asked").length;
  const answers = events.filter((e) => e.event === "answer_received").length;
  return questions === answers;
}

/**
 * Verify every answer_received has a matching analysis_written.
 */
export function allAnswersAnalyzed(events: TestLogEvent[]): boolean {
  const answers = events.filter((e) => e.event === "answer_received").length;
  const analyses = events.filter((e) => e.event === "analysis_written").length;
  return answers === analyses;
}

/**
 * Required frontmatter fields for transcript files.
 */
export const TRANSCRIPT_FIELDS = [
  "id",
  "title",
  "type",
  "created",
  "modified",
  "author",
  "summary",
  "project",
  "session",
  "specialist",
];

/**
 * Required frontmatter fields for analysis files.
 */
export const ANALYSIS_FIELDS = [
  "id",
  "title",
  "type",
  "created",
  "modified",
  "author",
  "summary",
  "related",
  "project",
  "session",
  "specialist",
];

File: testing/unit/harness/assertions.py
"""
Filesystem and interview-specific assertions.

Python reimplementation of tests/harness/lib/assertions.ts.
"""

import json
import os
import re
from pathlib import Path

# ── Filesystem assertions ────────────────────────────────────────────


def file_exists(base_dir: str, filename: str) -> bool:
    """Return True if the file exists under base_dir."""
    return Path(base_dir, filename).exists()


def file_contains(base_dir: str, filename: str, text: str) -> bool:
    """Return True if the file exists and contains the given text."""
    full_path = Path(base_dir, filename)
    if not full_path.exists():
        return False
    return text in full_path.read_text(encoding="utf-8")


def file_matches(base_dir: str, filename: str, pattern) -> bool:
    """Return True if the file exists and its content matches the compiled regex."""
    full_path = Path(base_dir, filename)
    if not full_path.exists():
        return False
    return bool(pattern.search(full_path.read_text(encoding="utf-8")))


def list_files(base_dir: str, subdir: str) -> list:
    """Return a sorted list of filenames (not dirs) in base_dir/subdir."""
    full_path = Path(base_dir, subdir)
    if not full_path.exists():
        return []
    return sorted(
        entry.name for entry in full_path.iterdir() if entry.is_file()
    )


# ── Interview-specific assertions ─────────────────────────────────────

# Timestamp filename pattern: YYYY-MM-DD-HH-MM-SS-slug.md
_TIMESTAMP_FILENAME = re.compile(r"^\d{4}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}-.+\.md$")


def all_files_match_timestamp_pattern(base_dir: str, subdir: str) -> bool:
    """Return True if all files in the directory match the timestamp naming convention."""
    files = list_files(base_dir, subdir)
    return len(files) > 0 and all(_TIMESTAMP_FILENAME.match(f) for f in files)


def has_valid_frontmatter(base_dir: str, filename: str, required_fields: list) -> bool:
    """Return True if the file has YAML frontmatter containing all required fields."""
    full_path = Path(base_dir, filename)
    content = full_path.read_text(encoding="utf-8")
    match = re.search(r"^---\n([\s\S]*?)\n---", content)
    if not match:
        return False
    frontmatter = match.group(1)
    return all(
        re.search(rf"^{re.escape(field)}:", frontmatter, re.MULTILINE)
        for field in required_fields
    )


def parse_test_log(base_dir: str, filename: str) -> list:
    """Parse a JSONL test log and return a list of event dicts."""
    full_path = Path(base_dir, filename)
    content = full_path.read_text(encoding="utf-8").strip()
    return [json.loads(line) for line in content.splitlines() if line.strip()]


def specialists_invoked(events: list) -> list:
    """Return unique specialist names in order of first appearance."""
    seen = []
    seen_set = set()
    for e in events:
        if e.get("event") == "specialist_invoked":
            name = e.get("specialist")
            if name and name not in seen_set:
                seen.append(name)
                seen_set.add(name)
    return seen


def exchange_count(events: list) -> int:
    """Count question_asked events."""
    return sum(1 for e in events if e.get("event") == "question_asked")


def all_questions_answered(events: list) -> bool:
    """Return True if every question_asked has a matching answer_received."""
    questions = sum(1 for e in events if e.get("event") == "question_asked")
    answers = sum(1 for e in events if e.get("event") == "answer_received")
    return questions == answers


def all_answers_analyzed(events: list) -> bool:
    """Return True if every answer_received has a matching analysis_written."""
    answers = sum(1 for e in events if e.get("event") == "answer_received")
    analyses = sum(1 for e in events if e.get("event") == "analysis_written")
    return answers == analyses


# ── Required frontmatter field lists ─────────────────────────────────

TRANSCRIPT_FIELDS = [
    "id",
    "title",
    "type",
    "created",
    "modified",
    "author",
    "summary",
    "project",
    "session",
    "specialist",
]

ANALYSIS_FIELDS = [
    "id",
    "title",
    "type",
    "created",
    "modified",
    "author",
    "summary",
    "related",
    "project",
    "session",
    "specialist",
]
